core/src/main/java/feast/core/service/StatsService.java

```diff
@@ -0,0 +1,587 @@
+    // Else, add to the list of features we still need to retrieve statistics for.
+    for (String featureName : features) {
+      FeatureReference featureReference =
+          new FeatureReference(
+              featureSetSpec.getProject(),
+              featureSetSpec.getName(),
+              featureSetSpec.getVersion(),
+              featureName);
+      Feature feature = Feature.withReference(featureReference);
+      Optional<FeatureStatistics> cachedFeatureStatistics = Optional.empty();
+      if (!forceRefresh) {
+        cachedFeatureStatistics =
+            featureStatisticsRepository.findFeatureStatisticsByFeatureAndDate(feature, date);
+      }
+      if (cachedFeatureStatistics.isPresent()) {
+        featureNameStatistics.add(cachedFeatureStatistics.get().toProto());
+      } else {
+        featuresMissingStats.add(featureName);
+      }
+    }
+
+    // Retrieve the balance of statistics after checking the cache, and add it to the
+    // list of FeatureNameStatistics.
+    if (featuresMissingStats.size() > 0) {
+      FeatureSetStatistics featureSetStatistics =
+          statisticsRetriever.getFeatureStatistics(
+              featureSetSpec,
+              featuresMissingStats,
+              Timestamp.newBuilder().setSeconds(timestamp).build());
+
+      // Persist the newly retrieved statistics in the cache.
+      for (FeatureNameStatistics stat : featureSetStatistics.getFeatureNameStatistics()) {
+        FeatureStatistics featureStatistics =
+            FeatureStatistics.fromProto(
+                featureSetSpec.getProject(),
+                featureSetSpec.getName(),
+                featureSetSpec.getVersion(),
+                stat,
+                date);
+        Optional<FeatureStatistics> existingRecord =
+            featureStatisticsRepository.findFeatureStatisticsByFeatureAndDate(
+                featureStatistics.getFeature(), date);
+        existingRecord.ifPresent(statistics -> featureStatistics.setId(statistics.getId()));
+        featureStatisticsRepository.save(featureStatistics);
+      }
+      featureNameStatistics.addAll(featureSetStatistics.getFeatureNameStatistics());
+    }
+    return featureNameStatistics;
+  }
+
+  /**
+   * Get the {@link StatisticsRetriever} corresponding to the store name provided.
+   *
+   * @param storeName name of the store to retrieve statistics from
+   * @return {@link StatisticsRetriever}
+   * @throws InvalidProtocolBufferException
+   */
+  private StatisticsRetriever getStatisticsRetriever(String storeName)
+      throws InvalidProtocolBufferException {
+    Store store = storeRepository.getOne(storeName).toProto();
+    if (store.getType() != StoreType.BIGQUERY) {
+      throw new IllegalArgumentException(
+          String.format(
+              "Invalid store %s specified. Batch statistics are only supported for BigQuery stores",
+              store.getType()));
+    }
+    return BigQueryStatisticsRetriever.newBuilder()
```

Comment:
Can we please use a static factory method to make this consistent with the other retrievers?

`fromConfig(Map<String, String> config)`?

Reply:
done